dev/tools/create_api_docs.dart

```diff
@@ -0,0 +1,1107 @@
+    final String metadata = template.replaceAll(
+      '{SITE_URL}',
+      branch == 'stable' ? 'https://api.flutter.dev/' : 'https://master-api.flutter.dev/',
+    );
+    metadataPath.parent.create(recursive: true);
+    metadataPath.writeAsStringSync(metadata);
+  }
+
+  Future<void> _createDocset() async {
+    // Must have dashing installed: go get -u github.com/technosophos/dashing
+    // Dashing produces a LOT of log output (~30MB), so we collect it, and just
+    // show the end of it if there was a problem.
+    print('${DateTime.now().toUtc()}: Building Flutter docset.');
+
+    // If dashing gets stuck, Cirrus will time out the build after an hour, and we
+    // never get to see the logs. Thus, we run it in the background and tail the
+    // logs only if it fails.
+    final ProcessWrapper result = ProcessWrapper(
+      await processManager.start(
+        <String>[
+          'dashing',
+          'build',
+          '--source',
+          publishRoot.path,
+          '--config',
+          docsRoot.childFile('dashing.json').path,
+        ],
+        workingDirectory: packageRoot.path,
+      ),
+    );
+    final List<int> buffer = <int>[];
+    result.stdout.listen(buffer.addAll);
+    result.stderr.listen(buffer.addAll);
+    // If the dashing process exited with an error, print the last 200 lines of stderr and exit.
+    final int exitCode = await result.done;
+    if (exitCode != 0) {
+      print('Dashing docset generation failed with code $exitCode');
+      final List<String> output = systemEncoding.decode(buffer).split('\n');
+      print(output.sublist(math.max(output.length - 200, 0)).join('\n'));
+      exit(exitCode);
+    }
+    buffer.clear();
+
+    // Copy the favicon file to the output directory.
+    final File faviconFile =
+        publishRoot.childDirectory('flutter').childDirectory('static-assets').childFile('favicon.png');
+    final File iconFile = packageRoot.childDirectory('flutter.docset').childFile('icon.png');
+    faviconFile..createSync(recursive: true)..copySync(iconFile.path);
+
+    // Post-process the dashing output.
+    final File infoPlist =
+        packageRoot.childDirectory('flutter.docset').childDirectory('Contents').childFile('Info.plist');
+    String contents = infoPlist.readAsStringSync();
+
+    // Since I didn't want to add the XML package as a dependency just for this,
+    // I just used a regular expression to make this simple change.
+    final RegExp findRe = RegExp(r'(\s*<key>DocSetPlatformFamily</key>\s*<string>)[^<]+(</string>)', multiLine: true);
+    contents = contents.replaceAllMapped(findRe, (Match match) {
+      return '${match.group(1)}dartlang${match.group(2)}';
+    });
+    infoPlist.writeAsStringSync(contents);
+    final Directory offlineDir = publishRoot.childDirectory('offline');
+    if (!offlineDir.existsSync()) {
+      offlineDir.createSync(recursive: true);
+    }
+    tarDirectory(packageRoot, offlineDir.childFile('flutter.docset.tar.gz'));
```

Comment:
Do we call this on Windows? I wouldn't expect `tar` to be present on Windows

Reply:
This is only expected to work on Linux and macOS. It is only called on Linux in CI.

Reply:
That would be another advantage of rewriting the bash script in Dart: it might run on Windows.  Of course, you'd still have to have `tar`, since that's what the expected output for the offline docs is, but you could use the `archive` package to do it.  It generates ZIP files that are about 40% larger though, so you wouldn't want to use `package:archive` for that.

Reply:
oh, i forgot we only build docs on linux, carry on

Reply:
I'm trying to deprecate package:archive, btw https://github.com/flutter/flutter/issues/115660